### ISSUE TITLE
Neo4j 3.5.2 `key not found` error 

### ISSUE THREAD

**Issue by causaly1** [neo4j/neo4j#12148]
After upgrading Neo4j to 3.5.1 (and then 3.5.2), we found that a query no longer works as expected (see "Steps to reproduce").  Instead of returning successfully, the following error is returned: `error: Neo4jError: key not found:   a@7`.

It is possible to avoid this error by replacing `sa` on line 11 with another name that is unique to the query.  The error can also be avoided by removing either: the `DISTINCT` keyword on line 10 , or `sa.f1 = "TEST" AND ` on line 26 of the query.  We saw nothing in any recent release notes to suggest that this behavior is intentional, and so believe that this might be a bug.

Edit:  We believe that the error is due to the fact that `sa` on line 11 can sometimes be `NULL`.

- **Neo4j version**: 3.5.2
- **Operating system**: Ubuntu 16.04
- **API/Driver**:  Cypher

### Steps to reproduce:
Execute the following Cypher query on a non-empty database:

```
MATCH (a:A)<-[:R1]-(d:B)<-[R2]-(c:C)<
-[:R3]-(e:E)
WHERE 
    NOT a:L1 
    AND NOT (c)<-[:R3]-(:F) 
    AND (a.f1 = "TEST") 
    
MATCH (a)-[:R4]->(g:G)-[:R1]->
(pg:G) 
WITH DISTINCT 
    a AS sa, 
    g AS sG, 
    pg AS spg,
    d AS d_c, 
    c, 
    e.f2 as f2
ORDER BY c.f3 DESC 
LIMIT 200000 

MATCH (sa)<-[:R1]-(d_c)<--(c)<-
[R2]-(sb:B)-[:R1]->(a:A)-[:R4]->
(g:G)-[:R1]->(pg:G) 
WHERE 
    NOT a:L1 
    AND (
        (sa.f1 = "TEST" AND g.f4 = "TEST2")) 

RETURN COUNT(*)
```
### Expected behavior
This query should return something

### Actual behavior
Query execution fails, and the following error is returned: `error: Neo4jError: key not found:   a@7`

**Comment by Lojjs** [comment]
@causaly1 This is most likely a bug, since we have seen a few similar ones introduced by the new index-backed-order-by feature in 3.5. However, I'm not able to reproduce you problem on an empty database in neither 3.5.1 nor 3.5.2 so it would be helpful if you could send over your data or an smaller example dataset where the problem occurs.

Best regards Louise, Neo4j Cypher team

**Comment by causaly1** [comment]
@Lojjs unfortunately we are not able to share this data with third parties.

**Comment by Hunterness** [comment]
Hi @causaly1, 

Is it possible for you to construct an example set of made up data where the problem occur? It would greatly help our investigation and enables you to not share any real data with us.

Best regards Therese, Neo4j Cypher team

**Comment by Lojjs** [comment]
@causaly1 Can you confirm that you have an index on `:A(f1)`? With such an index and a simple example graph with your labels and properties I seem to be able to reproduce the bug without any data from your side. Will investigate the cause and make a fix during this week.

**Comment by causaly1** [comment]
Hi,
@Hunterness we are also not able to provide our data model, however we can try providing a minimal working example if you would find that helpful.

@Lojjs you are correct.

**Comment by Lojjs** [comment]
Perfect, then you don't need to provide any more data to us at this point.

**Comment by Lojjs** [comment]
@causaly1 This turned out to be a bug with `DISTINCT` together with index-backed property lookup which was added in 3.5 (https://neo4j.com/docs/cypher-manual/current/query-tuning/cypher-index-values-order/ in case you are interrested in the details). A bug fix is now merged and will be included in the next 3.5 patch release (3.5.4). So from then on your query will not just work but hopefully also be able to take advantage of the performance increase index-backed property lookup gives. 

Best regards Louise
